Crashlytics/Crashlytics/Controllers/FIRCLSMetricKitManager.m

```diff
@@ -0,0 +1,320 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import <Foundation/Foundation.h>
+
+#import "Crashlytics/Crashlytics/Controllers/FIRCLSMetricKitManager.h"
+
+#if CLS_METRICKIT_SUPPORTED
+
+#import "Crashlytics/Crashlytics/Controllers/FIRCLSManagerData.h"
+#import "Crashlytics/Crashlytics/Helpers/FIRCLSFile.h"
+#import "Crashlytics/Crashlytics/Helpers/FIRCLSLogger.h"
+#import "Crashlytics/Crashlytics/Models/FIRCLSExecutionIdentifierModel.h"
+#import "Crashlytics/Crashlytics/Models/FIRCLSInternalReport.h"
+#import "Crashlytics/Crashlytics/Public/FirebaseCrashlytics/FIRCrashlyticsReport.h"
+
+@interface FIRCLSMetricKitManager ()
+
+@property FBLPromise *metricKitDataAvailable;
+@property FIRCLSExistingReportManager *existingReportManager;
+@property FIRCLSFileManager *fileManager;
+@property FIRCLSManagerData *managerData;
+
+@end
+
+@implementation FIRCLSMetricKitManager
+
+- (instancetype)initWithManagerData:(FIRCLSManagerData *)managerData
+              existingReportManager:(FIRCLSExistingReportManager *)existingReportManager
+                        fileManager:(FIRCLSFileManager *)fileManager {
+  _existingReportManager = existingReportManager;
+  _fileManager = fileManager;
+  _managerData = managerData;
+  return self;
+}
+
+/*
+ * Registers the MetricKit manager to receive MetricKit reports by adding self to the
+ * MXMetricManager subscribers. Also initializes the promise that we'll use to ensure that any
+ * MetricKit report files are included in Crashylytics fatal reports. If no crash occurred on the
+ * last run of the app, this promise is immediately resolved so that the upload of any nonfatal
+ * events can proceed.
+ */
+- (void)registerMetricKitManager {
+  [[MXMetricManager sharedManager] addSubscriber:self];
+  self.metricKitDataAvailable = [FBLPromise pendingPromise];
+
+  // If there was no crash on the last run of the app, then we aren't expecting a MetricKit
+  // diagnostic report and should resolve the promise immediately.
+  if (![self.fileManager didCrashOnPreviousExecution] && ![self.fileManager didCrashPreviously]) {
+    @synchronized(self) {
+      [self.metricKitDataAvailable fulfill:nil];
```

Comment:
So this is really tricky, but I'm worried about this block of code for 2 reasons:
 - If we don't detect the crash and don't write out the marker file, but MetricKit does, we won't capture the MetricKit crash, so we might miss out on events.
 - If we detect a crash by writing out the marker file, but MetricKit does NOT, then this doesn't look like it'll resolve, and we won't upload crash reports.

Reply:
If we don't detect the crash we would still receive the diagnostic payload from MetricKit, we just wouldn't send it up this run of the app (it would be treated as a non-fatal, in terms of timing). I'm not sure I understand what you mean in the second point though - the registerMetricKitManager method will be called whenever the app opens, regardless of if something crashed.